system/lib/wasmfs/paths.h

```diff
@@ -5,44 +5,71 @@
 
 #pragma once
 
-#include <vector>
+#include <cassert>
+#include <memory>
+#include <string_view>
+#include <variant>
 
 #include "file.h"
 
-namespace wasmfs {
+namespace wasmfs::path {
 
-struct ParsedPath {
-  std::optional<Directory::Handle> parent;
-  std::shared_ptr<File> child;
+// Typically -ENOTDIR or -ENOENT.
+using Error = long;
+
+// The parent directory and the name of an entry within it.
+using ParentChild = std::pair<std::shared_ptr<Directory>, std::string_view>;
```

Comment:
What holds the memory the `string_view` refers to? I'm not sure how to reason about this being safe and not referring to released memory.

If the directory were locked then I'd say referring to child names is valid. But we only have a reference here?

Reply:
The returned string view is typically a sub-slice of the input string view, which is in turn usually a view into the `const char*` string the user passed to the syscall. I'll add a comment.

Reply:
Not sure I follow. What is "original path" in the comment? The one passed to the syscall?

It seems reasonable to depend on that, but maybe it should be very explicit. In particular, `parse*` can return a view with the same lifetime of the input, but nothing else?

Reply:
I pushed a commit making the comment more specific now.

```
// The parent directory and the name of an entry within it. The returned string
// view is either backed by the same memory as the view passed to `parseParent`
// or is a view into a static string.
```

Reply:
Makes sense, thanks!